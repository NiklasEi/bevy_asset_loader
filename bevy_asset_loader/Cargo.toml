[package]
name = "bevy_asset_loader"
version = "0.20.1"
authors = ["Niklas Eicker <git@nikl.me>"]
edition = "2021"
license = "MIT OR Apache-2.0"
description = "Bevy plugin for asset loading"
repository = "https://github.com/NiklasEi/bevy_asset_loader"
homepage = "https://github.com/NiklasEi/bevy_asset_loader"
documentation = "https://docs.rs/bevy_asset_loader"
keywords = ["bevy", "gamedev", "asset", "assets"]
categories = ["game-development"]
rust-version = "1.74.0"
readme = "README.md"

[features]
# This feature adds support for bevy's TextureAtlas assets
2d = ["bevy/bevy_sprite", "bevy_asset_loader_derive/2d"]
# This feature adds support for bevy's StandardMaterial assets
3d = ["bevy/bevy_pbr", "bevy_asset_loader_derive/3d"]
standard_dynamic_assets = ["dep:bevy_common_assets", "dep:serde"]
progress_tracking = ["dep:iyes_progress"]

[dependencies]
bevy = { version = "0.13", default-features = false, features = ["bevy_asset"] }
bevy_asset_loader_derive = { version = "=0.20.1", path = "../bevy_asset_loader_derive" }
anyhow = "1"
path-slash = "0.2"

bevy_common_assets = { version = "0.10", features = ["ron"], optional = true }
serde = { version = "1", optional = true }
iyes_progress = { version = "0.11", optional = true }

[dev-dependencies]
bevy = { version = "0.13", features = ["vorbis"] }
anyhow = "1"
iyes_progress = { version = "0.11" }
bevy_common_assets = { version = "0.10", features = ["ron"] }
serde = { version = "1" }
ron = "0.8.1"
trybuild = { version = "1.0" }

[package.metadata.docs.rs]
all-features = true
rustdoc-args = ["--cfg", "docsrs"]

[lints]
workspace = true

[[example]]
name = "two_collections"
path = "examples/two_collections.rs"

[[example]]
name = "atlas_from_grid"
path = "examples/atlas_from_grid.rs"
required-features = ["2d"]

[[example]]
name = "init_resource"
path = "examples/init_resource.rs"

[[example]]
name = "manual_dynamic_asset"
path = "examples/manual_dynamic_asset.rs"
required-features = ["standard_dynamic_assets"]

[[example]]
name = "dynamic_asset"
path = "examples/dynamic_asset.rs"
required-features = ["2d", "standard_dynamic_assets"]

[[example]]
name = "no_loading_state"
path = "examples/no_loading_state.rs"
required-features = ["2d"]

[[example]]
name = "standard_material"
path = "examples/standard_material.rs"
required-features = ["3d"]

[[example]]
name = "progress_tracking"
path = "examples/progress_tracking.rs"
required-features = ["progress_tracking", "2d"]

[[example]]
name = "failure_state"
path = "examples/failure_state.rs"

[[example]]
name = "full_collection"
path = "examples/full_collection.rs"
required-features = ["2d", "3d"]

[[example]]
name = "full_dynamic_collection"
path = "examples/full_dynamic_collection.rs"
required-features = ["2d", "3d", "standard_dynamic_assets"]

[[example]]
name = "custom_dynamic_assets"
path = "examples/custom_dynamic_assets.rs"

[[example]]
name = "image_asset"
path = "examples/image_asset.rs"
required-features = ["2d"]

[[example]]
name = "configure_loading_state"
path = "examples/configure_loading_state.rs"
required-features = ["2d"]

[[example]]
name = "asset_maps"
path = "examples/asset_maps.rs"
