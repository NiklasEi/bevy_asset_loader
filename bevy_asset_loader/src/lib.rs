//! The goal of this crate is to offer an easy way for bevy games to load all their assets in a loading [`State`](::bevy::ecs::schedule::State).
//!
//! `bevy_asset_loader` introduces the derivable trait [`AssetCollection`](crate::asset_collection::AssetCollection). Structs with asset handles
//! can be automatically loaded during a configurable loading [`State`](::bevy::ecs::schedule::State). Afterwards they will be inserted as
//! resources containing loaded handles and the plugin will switch to a second configurable [`State`](::bevy::ecs::schedule::State).
//!
//! ```edition2021
//! # use bevy_asset_loader::prelude::*;
//! # use bevy::prelude::*;
//! # use bevy::asset::AssetPlugin;
//!
//! fn main() {
//!     App::new()
//! # /*
//!         .add_plugins(DefaultPlugins)
//! # */
//! #       .add_plugins((MinimalPlugins, AssetPlugin::default()))
//! #       .init_resource::<iyes_progress::ProgressCounter>()
//!         .init_state::<GameState>()
//!         .add_loading_state(
//!             LoadingState::new(GameState::Loading)
//!                 .continue_to_state(GameState::Next)
//!                 .load_collection::<AudioAssets>()
//!                 .load_collection::<ImageAssets>()
//!         )
//!         .add_systems(Update, use_asset_handles.run_if(in_state(GameState::Next)))
//! #       .set_runner(|mut app| {app.update(); AppExit::Success})
//!         .run();
//! }
//!
//! #[derive(AssetCollection, Resource)]
//! struct AudioAssets {
//!     #[asset(path = "audio/background.ogg")]
//!     background: Handle<AudioSource>,
//!     #[asset(path = "audio/plop.ogg")]
//!     plop: Handle<AudioSource>
//! }
//!
//! #[derive(AssetCollection, Resource)]
//! pub struct ImageAssets {
//!     #[asset(path = "images/player.png")]
//!     pub player: Handle<Image>,
//!     #[asset(path = "images/tree.png")]
//!     pub tree: Handle<Image>,
//! }
//!
//! // since this function runs in MyState::Next, we know our assets are loaded.
//! // We can get their handles from the AudioAssets resource.
//! fn use_asset_handles(mut commands: Commands, audio_assets: Res<AudioAssets>) {
//!     commands.spawn(AudioBundle {
//!         source: audio_assets.background.clone(),
//!         ..default()
//!     });
//! }
//!
//! #[derive(Clone, Eq, PartialEq, Debug, Hash, Default, States)]
//! enum GameState {
//!     #[default]
//!     Loading,
//!     Next
//! }
//! ```

#![forbid(unsafe_code)]
#![warn(unused_imports, missing_docs)]
#![cfg_attr(docsrs, feature(doc_cfg))]

/// Trait definition for types that represent a collection of assets
///
/// And extension traits to insert said collections into your Bevy app or world
pub mod asset_collection;
/// Types and infrastructure to load and use dynamic assets
pub mod dynamic_asset;
/// A game state responsible for loading assets
pub mod loading_state;
/// Trait definition for mapped assets collection
pub mod mapped;
/// Dynamic assets for common Bevy asset types
#[cfg_attr(docsrs, doc(cfg(feature = "standard_dynamic_assets")))]
#[cfg(feature = "standard_dynamic_assets")]
pub mod standard_dynamic_asset;

/// Most commonly used types
pub mod prelude {
    #[doc(hidden)]
    pub use crate::loading_state::config::{ConfigureLoadingState, LoadingStateConfig};
    #[doc(hidden)]
    #[cfg(feature = "standard_dynamic_assets")]
    pub use crate::standard_dynamic_asset::{
        RegisterStandardDynamicAsset, StandardDynamicAsset, StandardDynamicAssetArrayCollection,
        StandardDynamicAssetCollection,
    };
    #[doc(hidden)]
    pub use crate::{
        asset_collection::{AssetCollection, AssetCollectionApp, AssetCollectionWorld},
        dynamic_asset::{
            DynamicAsset, DynamicAssetCollection, DynamicAssetCollections, DynamicAssetType,
            DynamicAssets,
        },
        loading_state::{LoadingState, LoadingStateAppExt, LoadingStateSet},
        mapped::{AssetFileName, AssetFileStem, AssetLabel, MapKey},
    };
}

#[cfg(all(feature = "2d", feature = "3d"))]
#[doc = include_str!("../../README.md")]
#[cfg(doctest)]
struct ReadmeDoctests;
