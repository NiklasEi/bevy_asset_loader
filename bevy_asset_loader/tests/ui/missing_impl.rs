use bevy_asset_loader::prelude::*;
use bevy::prelude::*;

fn main() {
    App::new().init_collection::<Test>();
}

#[derive(Resource)]
struct Test {
    material: Handle<StandardMaterial>,
}
