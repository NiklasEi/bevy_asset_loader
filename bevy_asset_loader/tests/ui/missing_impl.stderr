error[E0277]: `Test` is not an `AssetCollection`
 --> tests/ui/missing_impl.rs:5:34
  |
5 |     App::new().init_collection::<Test>();
  |                ---------------   ^^^^ invalid `AssetCollection`
  |                |
  |                required by a bound introduced by this call
  |
  = help: the trait `AssetCollection` is not implemented for `Test`
  = note: consider annotating `Test` with `#[derive(AssetCollection)]`
note: required by a bound in `bevy_asset_loader::asset_collection::AssetCollectionApp::init_collection`
 --> src/asset_collection.rs
  |
  |     fn init_collection<A: AssetCollection>(&mut self) -> &mut Self;
  |                           ^^^^^^^^^^^^^^^ required by this bound in `AssetCollectionApp::init_collection`
